File: backend-node/recommend.js
// 智能推荐算法

/**
 * 计算财务健康度
 * @param {Object} car 车辆信息
 * @param {Object} profile 用户画像
 * @returns {number} 0-100分
 */
function calculateFinancialHealth(car, profile) {
  let score = 100;
  
  // 计算首付
  const downPayment = car.price_discount * 0.2 * 10000; // 20%首付,转换为元
  
  // 检查首付是否合理(不超过存款的50%)
  if (downPayment > profile.savings_for_down_payment * 0.5) {
    score -= 30;
  } else if (downPayment > profile.savings_for_down_payment * 0.3) {
    score -= 15;
  }
  
  // 计算月供(3年期,年利率3%)
  const loanAmount = car.price_discount * 0.8 * 10000;
  const monthlyRate = 0.03 / 12;
  const monthlyPayment = loanAmount * monthlyRate * Math.pow(1 + monthlyRate, 36) / 
                         (Math.pow(1 + monthlyRate, 36) - 1);
  
  // 检查月供是否合理(不超过月收入的30%)
  const income = profile.family_monthly_income || profile.monthly_income;
  if (monthlyPayment > income * 0.3) {
    score -= 25;
  } else if (monthlyPayment > income * 0.2) {
    score -= 10;
  }
  
  // 考虑其他负债
  if (profile.has_mortgage && profile.mortgage_monthly) {
    const totalPayment = monthlyPayment + profile.mortgage_monthly;
    if (totalPayment > income * 0.5) {
      score -= 20;
    }
  }
  
  return Math.max(0, score);
}

/**
 * 计算家庭适配度
 * @param {Object} car 车辆信息
 * @param {Object} profile 用户画像
 * @returns {number} 0-100分
 */
function calculateFamilyFit(car, profile) {
  let score = 100;
  
  // 家庭成员数量
  const totalMembers = profile.family_members + (profile.living_with_parents ? 2 : 0);
  
  // 座位需求
  if (profile.has_children) {
    if (car.seat_count < 5) {
      score -= 30;
    } else if (car.seat_count >= 7 && totalMembers >= 5) {
      score += 10; // 大家庭适合7座
    }
  }
  
  // 车身类型匹配
  if (totalMembers >= 5) {
    if (car.body_type === 'mpv' || car.body_type === 'suv') {
      score += 15;
    } else if (car.body_type === 'sedan') {
      score -= 10;
    }
  }
  
  // 三代同行
  if (profile.living_with_parents && profile.has_children) {
    if (car.seat_count >= 7) {
      score += 20;
    } else if (car.seat_count < 7) {
      score -= 15;
    }
  }
  
  return Math.max(0, Math.min(100, score));
}

/**
 * 计算场景匹配度
 * @param {Object} car 车辆信息
 * @param {Object} profile 用户画像
 * @returns {number} 0-100分
 */
function calculateScenarioMatch(car, profile) {
  let score = 50; // 基础分
  
  switch (profile.primary_usage) {
    case 'commute':
      // 通勤优先混动/纯电
      if (car.energy_type === 'bev' || car.energy_type === 'phev' || car.energy_type === 'hybrid') {
        score += 30;
      }
      // 考虑充电条件
      if (profile.parking_condition === 'fixed_with_charger' && car.energy_type === 'bev') {
        score += 15;
      }
      break;
      
    case 'family':
      // 家庭使用优先SUV/MPV
      if (car.body_type === 'suv' || car.body_type === 'mpv') {
        score += 30;
      }
      break;
      
    case 'business':
      // 商务优先轿车
      if (car.body_type === 'sedan') {
        score += 30;
      }
      // 品牌加成
      if (['宝马', '奔驰', '奥迪'].includes(car.brand)) {
        score += 15;
      }
      break;
      
    case 'travel':
      // 自驾游优先SUV
      if (car.body_type === 'suv') {
        score += 25;
      }
      // 越野车加成
      if (car.brand === '坦克') {
        score += 20;
      }
      break;
  }
  
  return Math.min(100, score);
}

/**
 * 计算首车系数
 * @param {Object} car 车辆信息
 * @param {Object} profile 用户画像
 * @returns {number} 0-100分
 */
function calculateFirstCarFactor(car, profile) {
  if (profile.car_ownership !== 'none') {
    return 50; // 非首车购买
  }
  
  let score = 80;
  
  // 首车优先性价比高的
  const pricePerSeat = car.price_discount / car.seat_count;
  if (pricePerSeat < 3) {
    score += 15;
  } else if (pricePerSeat > 8) {
    score -= 15;
  }
  
  // 首车优先易驾驶的
  if (car.body_type === 'sedan' || car.body_type === 'hatchback') {
    score += 10;
  }
  
  return Math.max(0, Math.min(100, score));
}

/**
 * 计算性价比
 * @param {Object} car 车辆信息
 * @returns {number} 0-100分
 */
function calculateValueScore(car) {
  let score = 50;
  
  // 优惠力度
  const discountRate = car.direct_discount / car.price_official;
  if (discountRate > 0.15) {
    score += 20;
  } else if (discountRate > 0.10) {
    score += 10;
  }
  
  // 贴息和置换补贴
  if (car.loan_subsidy_amount) {
    score += 10;
  }
  if (car.replacement_subsidy) {
    score += 10;
  }
  
  // 价格区间评分
  if (car.price_discount < 15) {
    score += 15; // 经济型
  } else if (car.price_discount < 30) {
    score += 10; // 中端
  }
  
  return Math.min(100, score);
}

/**
 * 计算政策红利
 * @param {Object} car 车辆信息
 * @param {Object} profile 用户画像
 * @returns {number} 0-100分
 */
function calculatePolicyBonus(car, profile) {
  let score = 50;
  
  // 新能源车在限牌城市
  if (profile.restricted_plate_city) {
    if (car.energy_type === 'bev' || car.energy_type === 'phev') {
      score += 30;
    }
  }
  
  // 置换补贴
  if (profile.old_car_trade_in && car.replacement_subsidy) {
    score += 20;
  }
  
  return Math.min(100, score);
}

/**
 * 生成推荐理由
 */
function generateReason(car, scores) {
  const reasons = [];
  
  if (scores.financial_health >= 80) {
    reasons.push('财务压力小');
  }
  if (scores.family_fit >= 80) {
    reasons.push('适合家庭使用');
  }
  if (scores.scenario_match >= 80) {
    reasons.push('匹配使用场景');
  }
  if (scores.value_score >= 80) {
    reasons.push('性价比高');
  }
  if (scores.policy_bonus >= 70) {
    reasons.push('享受政策红利');
  }
  
  return reasons.length > 0 ? reasons.join('，') : '综合推荐';
}

/**
 * 生成风险警告
 */
function generateRiskWarning(car, profile, scores) {
  const downPayment = car.price_discount * 0.2 * 10000;
  const loanAmount = car.price_discount * 0.8 * 10000;
  const monthlyRate = 0.03 / 12;
  const monthlyPayment = loanAmount * monthlyRate * Math.pow(1 + monthlyRate, 36) / 
                         (Math.pow(1 + monthlyRate, 36) - 1);
  
  const income = profile.family_monthly_income || profile.monthly_income;
  
  if (monthlyPayment > income * 0.5) {
    return {
      level: 'critical',
      message: `月供约${Math.round(monthlyPayment)}元，占收入的${Math.round(monthlyPayment / income * 100)}%`,
      suggestion: '建议选择更低价位的车型，或延长贷款期限'
    };
  } else if (monthlyPayment > income * 0.3) {
    return {
      level: 'high',
      message: `月供约${Math.round(monthlyPayment)}元，占收入的${Math.round(monthlyPayment / income * 100)}%`,
      suggestion: '月供略高，建议考虑财务风险'
    };
  } else if (downPayment > profile.savings_for_down_payment * 0.5) {
    return {
      level: 'medium',
      message: '首付占用存款比例较高',
      suggestion: '建议保留一定的应急储备金'
    };
  }
  
  return null;
}

/**
 * 推荐算法主函数
 */
function recommendCars(cars, profile, budgetMin, budgetMax) {
  const recommendations = cars
    .filter(car => {
      // 价格范围筛选
      if (budgetMin && car.price_discount < budgetMin) return false;
      if (budgetMax && car.price_discount > budgetMax) return false;
      return true;
    })
    .map(car => {
      // 计算各项评分
      const financial_health = calculateFinancialHealth(car, profile);
      const family_fit = calculateFamilyFit(car, profile);
      const scenario_match = calculateScenarioMatch(car, profile);
      const first_car_factor = calculateFirstCarFactor(car, profile);
      const value_score = calculateValueScore(car);
      const policy_bonus = calculatePolicyBonus(car, profile);
      
      // 综合评分(加权平均)
      const score = 
        financial_health * 0.30 +
        family_fit * 0.25 +
        scenario_match * 0.20 +
        first_car_factor * 0.10 +
        value_score * 0.10 +
        policy_bonus * 0.05;
      
      const scores = {
        financial_health,
        family_fit,
        scenario_match,
        first_car_factor,
        value_score,
        policy_bonus
      };
      
      return {
        car,
        score: Math.round(score),
        reason: generateReason(car, scores),
        ...scores,
        risk_warning: generateRiskWarning(car, profile, scores)
      };
    })
    .sort((a, b) => b.score - a.score)
    .slice(0, 10); // 返回前10个推荐
  
  return recommendations;
}

module.exports = { recommendCars };

File: backend/src/recommend.rs
// ── Recommendation Engine (User Profile × Cars) ───────────────
use carpricehub_shared::*;
use crate::db::Database;
use anyhow::Result;

pub async fn recommend(
    db: &Database,
    req: &RecommendRequest,
) -> Result<Vec<Recommendation>> {
    // Step 1: Build search query based on profile
    let mut query = SearchQuery {
        page: Some(1),
        page_size: Some(50),
        ..Default::default()
    };

    // Budget from profile if not specified
    if query.budget_min.is_none() {
        query.budget_min = req.budget_min;
    }
    if query.budget_max.is_none() {
        query.budget_max = req.budget_max;
    }

    // If still no budget, calculate from income (3x annual income as max)
    if query.budget_max.is_none() {
        let annual_income = req.profile.family_monthly_income.unwrap_or(req.profile.monthly_income) * 12;
        query.budget_max = Some((annual_income / 3) as f64);
    }

    // Energy type preference: prefer EV if restricted plate city
    if req.profile.restricted_plate_city {
        query.energy_type = Some(EnergyType::Bev);
    }

    // Seat count preference: 5+ if married with kids
    if req.profile.marital_status == MaritalStatus::Married && req.profile.has_children {
        query.seat_count_min = Some(5);
    }

    // Step 2: Search cars
    let cars = db.search_cars(&query).await?;

    // Step 3: Score each car
    let mut recommendations: Vec<Recommendation> = cars
        .into_iter()
        .map(|car| score_car(&car, &req.profile))
        .collect();

    // Step 4: Sort by total score
    recommendations.sort_by(|a, b| b.score.partial_cmp(&a.score).unwrap());

    // Step 5: Keep top 10
    recommendations.truncate(10);

    Ok(recommendations)
}

fn score_car(car: &Car, profile: &UserProfile) -> Recommendation {
    // 1. Financial Health (30%)
    let financial_health = calculate_financial_health(car, profile);

    // 2. Family Fit (25%)
    let family_fit = calculate_family_fit(car, profile);

    // 3. Scenario Match (20%)
    let scenario_match = calculate_scenario_match(car, profile);

    // 4. First Car Factor (10%)
    let first_car_factor = calculate_first_car_factor(car, profile);

    // 5. Value Score (10%)
    let value_score = calculate_value_score(car);

    // 6. Policy Bonus (5%)
    let policy_bonus = calculate_policy_bonus(car, profile);

    // Total score (0-100)
    let score = (financial_health * 0.30)
        + (family_fit * 0.25)
        + (scenario_match * 0.20)
        + (first_car_factor * 0.10)
        + (value_score * 0.10)
        + (policy_bonus * 0.05);

    // Risk warning
    let risk_warning = check_risk(car, profile);

    // Reason generation
    let reason = generate_reason(car, profile, score);

    Recommendation {
        car: car.clone(),
        score,
        reason,
        financial_health,
        family_fit,
        scenario_match,
        first_car_factor,
        value_score,
        policy_bonus,
        risk_warning,
    }
}

fn calculate_financial_health(car: &Car, profile: &UserProfile) -> f64 {
    let monthly_income = profile.family_monthly_income.unwrap_or(profile.monthly_income) as f64;
    let max_monthly_payment = profile.max_monthly_payment as f64;

    // Assume 30% down, 3-year loan, 5% interest (simplified)
    let loan_amount = (car.price_discount * 10000.0) * 0.7;
    let monthly_payment = loan_amount / 36.0 * 1.1; // rough estimate with interest

    if monthly_payment > max_monthly_payment {
        return 0.0; // Over budget
    }

    let payment_ratio = monthly_payment / monthly_income;
    let health_score = (1.0 - payment_ratio) * 100.0;

    health_score.max(0.0).min(100.0)
}

fn calculate_family_fit(car: &Car, profile: &UserProfile) -> f64 {
    if !profile.has_children && !profile.living_with_parents {
        // Single / couple: no strict requirement
        return 100.0;
    }

    let seats = car.seat_count.unwrap_or(5);
    let target_seats = if profile.family_members >= 5 { 7 } else { 5 };

    let seat_score = if seats >= target_seats { 100.0 } else { 50.0 };

    // ISOFIX presence (simplified: assume larger cars have it)
    let isofix_score = if car.body_type == BodyType::Suv || car.body_type == BodyType::Mpv { 100.0 } else { 70.0 };

    (seat_score + isofix_score) / 2.0
}

fn calculate_scenario_match(car: &Car, profile: &UserProfile) -> f64 {
    match profile.primary_usage {
        UsageScenario::Commute => {
            // Prefer fuel-efficient or EV for long commute
            if car.energy_type == EnergyType::Bev || car.energy_type == EnergyType::Hybrid {
                100.0
            } else if profile.daily_mileage > 50 {
                60.0 // High mileage petrol is expensive
            } else {
                85.0
            }
        }
        UsageScenario::Family => {
            // Prefer SUV/MPV with space
            if car.body_type == BodyType::Suv || car.body_type == BodyType::Mpv {
                100.0
            } else {
                60.0
            }
        }
        UsageScenario::Business => {
            // Prefer sedan/coupe for professional image
            if car.body_type == BodyType::Sedan || car.body_type == BodyType::Coupe {
                100.0
            } else {
                70.0
            }
        }
        UsageScenario::Travel => {
            // Prefer SUV for off-road / long trip
            if car.body_type == BodyType::Suv || car.body_type == BodyType::Pickup {
                100.0
            } else {
                65.0
            }
        }
    }
}

fn calculate_first_car_factor(car: &Car, profile: &UserProfile) -> f64 {
    if profile.car_ownership != CarOwnership::None {
        return 100.0; // Not first car, no preference
    }

    // First car: prefer reliable, easy to drive, cheap to insure
    // Simplified: smaller cars, lower price
    if car.price_discount < 15.0 {
        100.0
    } else if car.price_discount < 25.0 {
        80.0
    } else {
        60.0 // Premium cars less ideal for first car
    }
}

fn calculate_value_score(car: &Car) -> f64 {
    // Based on discount ratio and loan subsidy
    let discount_ratio = car.direct_discount / car.price_official;
    let subsidy_bonus = car.loan_subsidy_amount.unwrap_or(0.0) / car.price_official;

    let raw_score = (discount_ratio * 200.0) + (subsidy_bonus * 300.0);
    raw_score.max(0.0).min(100.0)
}

fn calculate_policy_bonus(car: &Car, profile: &UserProfile) -> f64 {
    let mut score = 50.0;

    // Restriction city bonus for EV
    if profile.restricted_plate_city && car.energy_type == EnergyType::Bev {
        score += 40.0;
    }

    // Trade-in subsidy
    if profile.old_car_trade_in && car.replacement_subsidy.is_some() {
        score += 20.0;
    }

    score.min(100.0)
}

fn check_risk(car: &Car, profile: &UserProfile) -> Option<RiskWarning> {
    let monthly_income = profile.family_monthly_income.unwrap_or(profile.monthly_income) as f64;
    let loan_amount = (car.price_discount * 10000.0) * 0.7;
    let monthly_payment = loan_amount / 36.0 * 1.1;

    let payment_ratio = monthly_payment / monthly_income;

    if payment_ratio > 0.50 {
        Some(RiskWarning {
            level: RiskLevel::Critical,
            message: "月供超过月收入50%，财务风险极高".to_string(),
            suggestion: "建议降低预算或增加首付".to_string(),
        })
    } else if payment_ratio > 0.40 {
        Some(RiskWarning {
            level: RiskLevel::High,
            message: "月供接近月收入40%，财务压力较大".to_string(),
            suggestion: "建议谨慎评估还款能力".to_string(),
        })
    } else if payment_ratio > 0.30 {
        Some(RiskWarning {
            level: RiskLevel::Medium,
            message: "月供超过推荐比例30%，需预留应急资金".to_string(),
            suggestion: "建议增加储蓄缓冲".to_string(),
        })
    } else {
        None
    }
}

fn generate_reason(car: &Car, profile: &UserProfile, score: f64) -> String {
    let mut reasons = vec![];

    if car.direct_discount > 0 {
        reasons.push(format!("当前优惠{}万", car.direct_discount));
    }

    if let Some(subsidy) = car.loan_subsidy_amount {
        if subsidy > 0 {
            reasons.push(format!("贴息{}万", subsidy));
        }
    }

    if profile.restricted_plate_city && car.energy_type == EnergyType::Bev {
        reasons.push("绿牌免摇号".to_string());
    }

    if profile.has_children && car.seat_count.unwrap_or(5) >= 5 {
        reasons.push("空间满足家庭需求".to_string());
    }

    if profile.car_ownership == CarOwnership::None && car.price_discount < 15.0 {
        reasons.push("适合首购".to_string());
    }

    if reasons.is_empty() {
        "综合性价比优秀".to_string()
    } else {
        format!("{}，综合评分{:.1}", reasons.join(" + "), score)
    }
}
